Added basic geometry collection

psavery opened the pull request:
Fixes: #94

psavery commented:
This is for issue #94.

Here is a basic geometry collection that I have made so far. What do you think of the structure?

As issue #94 mentions, I added `moleculeId` as a required index, which links to a parent molecule. I also have `cjson` as a required index, in which the geometry will be provided.

`provenanceType` would be the type of origin. Maybe it could be `calculation`, `user`, `web`, etc. The `provenanceId` could then be the `calculationId`, `userId`, `website`, respectively.

psavery commented:
I'm guessing I should also add some rest endpoints to get geometries?

alesgenova commented:
It's looking good so far at first glance!

There are some things we need to think about and decide on in my opinion:
- What will the relation between geometry and calculation be? I assume we will have a `geometryId` field on the calculation. Also the calculation already has a `cjson` field that contains the output of the calculation in term of geometry orbitals, vibrations, properties, etc. So there is an overlap in the `atoms` and `bonds` fields between the geometry and calculation cjsons. Do we duplicate the information, or do we do something else? Also, we should probably have an input and output geometry for calculations, just in case (they may refer to the same geometry).
- How is the geometry going to fit in the current workflow in jupyterlab? How are we going to search for existing geometries? If a user provides a new geometry are we gonna try and match it to existing geometries? (for instance to return an existing calculation instead of rerunning it from scratch)

Just some free thoughts I had on the new geometry model. This will add some needed flexibility, but I think we should all sit down and make a formal plan on how we will implement these relationships.

What do you think?

@psavery @cryos @cjh1

cryos commented:
The [meta tracking issue](https://github.com/OpenChemistry/42/issues/5) is a better place to discuss these things. I already addressed some of these things there.

psavery commented:
I made the provenance information optional, although I do believe it should be provided as often as possible, since we may not know the accuracy of a geometry without knowing its origin.

psavery commented:
@cjh1 @cryos Should I add some validation for the cjson a geometry takes? Maybe I should also make sure that the chemical formula in the cjson matches that of the parent molecule?

I can also whitelist the cjson to only take "atoms", "bonds", and "chemical json".

cjh1 commented:
@psavery For now I don't think we want to add to much validation for the cjson geometry, if and when use have a schema we should use that.

cjh1 commented:
@psavery The basic model is looking good ( may need to be updated for Girder 3? ). We should now move onto looking at now this model, for example the calculations  model.

psavery commented:
So, this is working now, and it is ready to be reviewed. You can currently:

1. Create a geometry that is linked to a parent molecule.
2. List geometries of a parent molecule.
3. Delete a geometry.

The geometry requires 2 things to be constructed: a parent molecule ID and cjson.

Here are some possible things we can add to this work, although I am not sure if they should be done in this PR or not:

1. When a molecule is created, also create a single geometry that links to the molecule. This geometry would have a duplicate cjson of the cjson contained in the molecule.
2. When a molecule is deleted, delete all of its geometries.
3. For running a calculation, allow a geometry to be used instead of a molecule.
4. If a user tries to upload a molecule that already exists (and they try to use a 3D format), automatically create a geometry with the 3D coordinates they used.

We don't really have any kind of validation to check a geometry a user tries to upload. Perhaps some validation would be: checking if the formula matches the parent molecule, checking if the connectivity matches, checking if another geometry exists that is identical.

cryos commented:
I don't think we need to create a geometry duplicate when a molecule is added. The major goal is to enable me to upload a geometry (which I think works in principle), and then to be able to specify that geometry in a calculation. We also need to consider how we can create a geometry from a geometry optimization in Open Babel, ANI, or a quantum code and trace back lineage if they have entries in a calculation object. Another one to consider is conformer generation, and then using conformer lists to do a number of single point energy calculations.

psavery commented:
@cryos That sounds good. The geometries have "provenanceType" and "provenanceId" keys. I think for geometry optimizations, we could automatically create a geometry from the results, and set the "provenanceType" to be "calculation" and the "provenanceId" to be the calculation Id, or something similar.

psavery commented:
@cjh1 I moved the geometry endpoints to the "/molecules" endpoint.

I think this is ready for review again. I added the ability to specify the geometry in the calculation so that a calculation may use a specific geometry instead of the default geometry (in the molecule).

cjh1 commented on girder/molecules/molecules/models/geometry.py:
```diff
@@ -0,0 +1,54 @@
+            geometry['provenanceType'] = provenanceType
+
+        if provenanceId is not None:
+            geometry['provenanceId'] = provenanceId
```
What values can `provenanceType` take? Is it free formed or should it be in forced as an enumeration?

psavery replied:
It is currently just a free formed string, where I was thinking to have some specific keywords such as "calculation", where we would expect the "provenanceId" to be the calculation id.

But we can limit the options in an enum if that would be preferable.

cjh1 replied:
Ok, that  makes sense we can keep it free formed for now, we may want it as an enumeration in the future, so the data is a little more structured.

psavery replied:
Sounds good.

Another keyword I was thinking of using would be something like "openbabel conformer". That one might not have a calculation to link back to (unless we run it as a task), but maybe the ID can be indices for the different conformers.

psavery commented:
Here's one design decision it might be good to get feedback on: I'm currently whitelisting the cjson that the user gives the server when a geometry is created. The whitelist only keeps the 'atoms', 'bonds', and 'chemicalJson' keys. Should I do this? Or not? The main reason I was doing it is because, when an optimization calculation finishes, it creates a new geometry, and if we did not whitelist the cjson, the geometry would contain a lot of other data including the molecular orbital data. I think that we can already get this data by linking back to the calculation where the geometry came from.

Also, FYI, the energy for a geometry that came from a calculation is currently being extracted server-side [here](https://github.com/OpenChemistry/mongochemserver/pull/112/files#diff-15b30d94c3467791b075f30c9872c344R49), which just parses the cjson keys `['properties']['totalEnergy']` to see if they exist. If we would rather the geometry model get the energy another way, such as a user providing it, let me know. It is good to have the energy in the top level of the geometry model, though, so that users can more easily do things like sort by energy.
